core/secp256k1_zkp: add blinding and unblinding API

romanz opened the pull request:
Notes to reviewers:
- The current code requires significant stack size increase (to ~40kB), so it depends on https://github.com/trezor/trezor-firmware/issues/174 to run on the TREZOR T.
- The tests can run on the emulator via:
```
$ make build_unix
$ ./tests/run_tests.sh test_trezor.crypto.curve.secp256k1.py
```
- I wasn't sure about the default values for `exp`, `min_bits` and `min_value` parameters. Should they be set by the user?

romanz committed:
core/secp256k1_zkp: add blinding and unblinding API

Allows rangeproof generation and rewinding, surjection proof generaion
and Pedersen commitments' balancing on the device.

NOTE: current code needs a lot of stack for rangeproofs (~40kB),
so it depends on trezor#174 (or just increasing the stack size via
`memory_T.ld`) to run on the TREZOR T.

real-or-random commented on core/embed/extmod/modtrezorcrypto/modtrezorcrypto-secp256k1_zkp.h:
```diff
@@ -325,13 +331,265 @@ STATIC mp_obj_t mod_trezorcrypto_secp256k1_context_multiply(
   }
+  // Verify and re-serialize into (optionally compressed) public key.
+  size_t outlen = sizeof(out);
+  if (!secp256k1_ec_pubkey_parse(ctx, &ec_pk, out, outlen)) {
```
Hm, somewhat annoying that we need to parse again. Wouldn't be cleaner to do in this hash callback? I think then also we don't need to do anything in the case of uncompressed output and can handle the compression on our own if necessary. (The group element output by `_ecdh` must be valid, so the check done by `_parse` is not necessary).

romanz replied:
Sounds good - fixed at https://github.com/romanz/trezor-firmware/pull/1/commits/729918821371de7462160ddc5e18f55944e26871.

real-or-random commented on core/embed/extmod/modtrezorcrypto/modtrezorcrypto-secp256k1_zkp.h:
```diff
@@ -298,17 +299,22 @@ static int secp256k1_ecdh_hash_passthrough(uint8_t *output, const uint8_t *x,
-/// def multiply(self, secret_key: bytes, public_key: bytes) -> bytes:
+/// def multiply(
+///     self, secret_key: bytes, public_key: bytes, compressed: bool = False
+/// ) -> bytes:
```
Would be good to document that `compressed` refers to the output and not the input (and the input can be compressed and uncompressed).

romanz replied:
Good suggestion, fixed at https://github.com/romanz/trezor-firmware/pull/1/commits/729918821371de7462160ddc5e18f55944e26871

real-or-random commented on core/embed/extmod/modtrezorcrypto/modtrezorcrypto-secp256k1_zkp.h:
```diff
@@ -325,13 +331,265 @@ STATIC mp_obj_t mod_trezorcrypto_secp256k1_context_multiply(
+    mp_obj_int_to_bytes_impl(o, true, 8, buff);
+    for (int i = 0; i < 8; i++) {
+      res <<= i > 0 ? 8 : 0;
+      res |= (uint64_t)(buff[i] & 0xff);
```
I wonder whether this is endian-dependent (if yes, should maybe be documented or #error out if wrong)

romanz replied:
IIUC `mp_obj_int_to_bytes_impl` should order the bytes according to its 2nd argument:
https://github.com/micropython/micropython/blob/b9ec6037edf5e6ff6f8f400d70f7351d1b0af67d/py/objint.h#L58
I have updated the invocation at https://github.com/romanz/trezor-firmware/pull/1/commits/729918821371de7462160ddc5e18f55944e26871 to make it more explicit.

romanz replied:
Refactored at b185e8475d22f05531eca0ec410d9cbd26a333e5.

real-or-random commented on core/embed/extmod/modtrezorcrypto/modtrezorcrypto-secp256k1_zkp.h:
```diff
@@ -325,13 +331,265 @@ STATIC mp_obj_t mod_trezorcrypto_secp256k1_context_multiply(
+    mp_raise_ValueError("Pedersen commit failed");
+  }
+
+  byte output[33] = {0};
```
nit: no need to initialize to 0. (don't know what the code style is)

romanz replied:
Removed the initialization.

real-or-random commented on core/embed/extmod/modtrezorcrypto/modtrezorcrypto-secp256k1_zkp.h:
```diff
@@ -325,13 +331,265 @@ STATIC mp_obj_t mod_trezorcrypto_secp256k1_context_multiply(
+
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(
+    mod_trezorcrypto_secp256k1_context_pedersen_commit_obj, 4, 4,
+    mod_trezorcrypto_secp256k1_context_pedersen_commit);
```
Is there a reason to write the function in the generic vararg way?

romanz replied:
Unfortunately, there is no `MP_DECLARE_CONST_FUN_OBJ_4` (and higher):
https://github.com/micropython/micropython/blob/b9ec6037edf5e6ff6f8f400d70f7351d1b0af67d/py/obj.h#L265-L271

real-or-random commented on core/embed/extmod/modtrezorcrypto/modtrezorcrypto-secp256k1_zkp.h:
```diff
@@ -325,13 +331,265 @@ STATIC mp_obj_t mod_trezorcrypto_secp256k1_context_multiply(
+///     '''
+STATIC mp_obj_t mod_trezorcrypto_secp256k1_context_rangeproof_sign(
+    size_t n_args, const mp_obj_t *args) {
+  // TODO: not sure about the constants values here.
```
Hm, yep, I'm not sure either, see
https://github.com/ElementsProject/elements/blob/7b302eb99d2d1bf861429d6bd209771f634b8e8e/src/blind.cpp#L184
 @apoelstra @gsansders  Can you comment?

real-or-random replied:
@instagibbs

romanz replied:
I changed the default `BITS` value to 36 (similar to Elements' defaults).

romanz replied:
926abd29813bbd4617c5117a689abdf92e9b2139 allows setting the range-proof configuration from MicroPython.

real-or-random commented on core/embed/extmod/modtrezorcrypto/modtrezorcrypto-secp256k1_zkp.h:
```diff
@@ -325,13 +331,265 @@ STATIC mp_obj_t mod_trezorcrypto_secp256k1_context_multiply(
+    mp_raise_ValueError("Generator parsing failed");
+  }
+
+  mp_buffer_info_t scratch_buffer;
```
nit: `scratch` is somewhat confusing because "scratch spaces" are an existing concept in secp256k1

romanz replied:
Changed to `proof_buffer` at https://github.com/romanz/trezor-firmware/pull/1/commits/f0ef2724ae5100cb7c797fb2ffde425d40be563f

real-or-random commented on core/embed/extmod/modtrezorcrypto/modtrezorcrypto-secp256k1_zkp.h:
```diff
@@ -325,13 +331,265 @@ STATIC mp_obj_t mod_trezorcrypto_secp256k1_context_multiply(
+  return mp_obj_new_memoryview('B', rangeproof_len, scratch_buffer.buf);
+}
+
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(
```
(same question here as above and further below, I'll won't mark additional instances)

romanz replied:
MicroPython doesn't seem to have `MP_DEFINE_CONST_FUN_OBJ_9` :(

real-or-random commented on core/embed/extmod/modtrezorcrypto/modtrezorcrypto-secp256k1_zkp.h:
```diff
@@ -325,13 +331,265 @@ STATIC mp_obj_t mod_trezorcrypto_secp256k1_context_multiply(
+  mp_get_buffer_raise(args[6], &asset_message, MP_BUFFER_WRITE);
+
+  size_t asset_message_len = asset_message.len;
+  byte value_blind[32] = {0};
```
(no need to init, also no need in the following three uint64_t declarations)

romanz replied:
Fixed at https://github.com/romanz/trezor-firmware/pull/1/commits/729918821371de7462160ddc5e18f55944e26871.

real-or-random commented on core/embed/extmod/modtrezorcrypto/modtrezorcrypto-secp256k1_zkp.h:
```diff
@@ -325,13 +331,265 @@ STATIC mp_obj_t mod_trezorcrypto_secp256k1_context_multiply(
+///                       nonce: bytes, range_proof: bytes,
+///                       extra_commit: bytes, asset_message: bytearray) ->
+///                       (value: long, blind: bytes,
+///                        asset_message: memoryview):
```
nit: the "asset" in `asset_message" is maybe too specific. This function is generic and messages could be other things.

romanz replied:
You're right, thanks!
Fixed at https://github.com/romanz/trezor-firmware/pull/1/commits/729918821371de7462160ddc5e18f55944e26871

real-or-random commented on core/embed/extmod/modtrezorcrypto/modtrezorcrypto-secp256k1_zkp.h:
```diff
@@ -325,13 +331,265 @@ STATIC mp_obj_t mod_trezorcrypto_secp256k1_context_multiply(
+  if (!secp256k1_rangeproof_rewind(
+          ctx, value_blind, &value, asset_message.buf, &asset_message_len,
+          nonce.buf, &min_value, &max_value, &commitment, range_proof.buf,
+          range_proof.len, extra_commit.buf, extra_commit.len, &generator)) {
```
This relies on the undocumented fact that `secp256k1_rangeproof_rewind` won't write more bytes than specified by it's 5th argument. https://github.com/ElementsProject/secp256k1-zkp/blob/secp256k1-zkp/src/modules/rangeproof/rangeproof_impl.h#L468
This is fine but we should add this guarantee to the upstream API (and maybe actually fail if the provided buffer was not large enough to hold the message instead of silently truncating the message without giving the caller a way to tell if truncation happened or not)

real-or-random replied:
> (and maybe actually fail if the provided buffer was not large enough to hold the message instead of silently truncating the message without giving the caller a way to tell if truncation happened or not)

Not sure if this part makes sense. Now that I look at it again, I think the message length cannot be determined by the rewinder (i.e., you cannot distinguish between a 96 bytes message and a 4096 byte message where the last 4000 bytes are 0x00). We'll need to look at this. 

Related: https://github.com/ElementsProject/secp256k1-zkp/issues/42

romanz replied:
We can re-use the pre-allocated `proof_buffer` (which is currently >5KB) for the message during the rewinding (currently it's only used for range/surjection proof generation): 9ae59f8e.

EDIT: I need to test if it works on the device (not sure if we'll have enough memory...)

romanz replied:
Unfortunately, since we need to hold both the user-provided rangeproof and the message buffer (for the `asset_message`) in memory in addition to the `secp256k1_context`, the unblinding start to fail with OOM errors.
In our case, we know that message should be 64 bytes - so IIUC it should be fine to "stop" the parsing (and truncate the rest of the message if it's longer), since we can re-blind the output to make sure we've got the correct asset and asset blind factor.

romanz replied:
>  the unblinding start to fail with OOM errors.

After a few memory optimizations, I think that we would be able fit the `proof_buffer` into RAM during unblinding :)
Will squash and rebase the PR with the fixes tomorrow.

romanz replied:
Done @ https://github.com/romanz/trezor-firmware/commit/1dd6e8972fb6854da1e9fdcedb08ef491d148330

romanz replied:
Update:
(Probably) due to heap fragmentation, large buffer allocation fail when ran after address/signature-related API invocations :(
I have added a compile flag at 7144d566 for allocating the largest buffer (containing the range-proof) statically at CCMRAM - which can be disabled when the stack would be moved there (https://github.com/trezor/trezor-firmware/issues/174).

real-or-random commented on core/embed/extmod/modtrezorcrypto/modtrezorcrypto-secp256k1_zkp.h:
```diff
@@ -364,11 +630,27 @@ STATIC const mp_obj_type_t mod_trezorcrypto_secp256k1_context_type = {
+///     '''
+///     Allocate a buffer, large enough for holding a range/surjection proof.
+///     '''
+STATIC mp_obj_t mod_trezorcrypto_secp256k1_zkp_allocate_scratch_buffer() {
```
(same here for "scratch")

romanz replied:
Fixed at https://github.com/romanz/trezor-firmware/pull/1/commits/f0ef2724ae5100cb7c797fb2ffde425d40be563f

real-or-random commented on core/embed/extmod/modtrezorcrypto/modtrezorcrypto-secp256k1_zkp.h:
```diff
@@ -325,13 +331,265 @@ STATIC mp_obj_t mod_trezorcrypto_secp256k1_context_multiply(
   }
+  // Verify and re-serialize into (optionally compressed) public key.
+  size_t outlen = sizeof(out);
+  if (!secp256k1_ec_pubkey_parse(ctx, &ec_pk, out, outlen)) {
```
"shared public key" sounds like this a public then. Maybe call it "shared secret group element"  instead.

romanz replied:
After removing the parsing at https://github.com/romanz/trezor-firmware/pull/1/commits/729918821371de7462160ddc5e18f55944e26871, the error message is not needed.

real-or-random reviewed (commented):
I looked just a the first commit. Looks good overall!

romanz commented:
Many thanks for the review!
Will fix the issues in the following days.

real-or-random commented on core/embed/extmod/modtrezorcrypto/modtrezorcrypto-secp256k1_zkp.h:
```diff
@@ -25,6 +25,23 @@
+// secp256k1_surjectionproof struct.
+#define SURJECTIONPROOF_MAX_USED_INPUTS 16
+#else
+// For emulator - to allocate >8kB secp256k1_surjectionproof struct.
```
I'm probably just not seeing the obvious but why do we want different code on the emulator?

romanz replied:
I wanted to support the verification of surjection proofs on the emulator (since we disabled it on the device, to reduce stack usage at https://github.com/ElementsProject/secp256k1-zkp/pull/57):
https://github.com/romanz/trezor-firmware/pull/1/files#diff-1aaa7015e3c406adc561e688e3404305R199
Not sure that it's the best way though... WDYT?

romanz replied:
Update: I removed the different definitions between emulator and device, since we now allocate the full surjection proof on CCMRAM (https://github.com/romanz/trezor-firmware/commit/7144d566e31be8986726bd5051ef67ca3db532a8).

romanz commented:
Thanks a lot for the review!
Please take a look at the updated code.

romanz commented:
Squashed, rebased over https://github.com/trezor/trezor-firmware/commits/master and force-pushed: https://github.com/romanz/trezor-firmware/pull/1/commits/1dd6e8972fb6854da1e9fdcedb08ef491d148330

romanz commented:
The commit above (https://github.com/romanz/trezor-firmware/pull/1/commits/7144d566e31be8986726bd5051ef67ca3db532a8) uses static allocation on CCMRAM for `proof_buffer` to prevent OOM errors.
When the stack moves to CCMRAM, we can disable this feature.

romanz commented:
This branch can be tested on the Model T device by running:
```
$ make build_firmware upload
$ sudo cat /dev/ttyACM0 | tee debug.log &
$ py.test -k "liquid or elements" -v
```

romanz commented:
Closing - will re-open to upstream.
